aggregator/src/app/AggregationStrategy.ts

```diff
@@ -2,216 +2,431 @@ import {
+    // because the bundles will be considered again in the same order.
+    //
+    // To fix this, we simply mark half the bundles as failed. We could
+    // isolate the issue by doing a lot of in-order reprocessing, but
+    // having this defense in place should prevent the attack in the first
+    // place, so false-positives here are a minor concern (keeping in mind
+    // these bundles will still get retried later).
+
+    let failureSample = shuffled(includedRows);
+    failureSample = failureSample.slice(0, failureSample.length / 2);
+
+    for (const row of failureSample) {
+      row.submitError = "Included in failure sample for unprofitable bundle";
     }
 
-    const feeRequired = await this.#measureRequiredFee(bundle);
+    failedRows.push(...failureSample);
 
-    const successes = bundleResult.returnValue.successes;
+    return {
+      aggregateBundle: nil,
+      includedRows: [],
+      bundleOverheadCost: result.bundleOverheadCost,
+      expectedFee: BigNumber.from(0),
+      expectedMaxCost: BigNumber.from(0),
+      failedRows,
+    };
+  }
+
+  async estimateFee(bundle: Bundle, bundleOverheadGas?: BigNumber) {
+    const [{ operationStatuses: successes, fee: feeDetected, errorReason }] =
+      await this
+        .#measureFees([
+          bundle,
+        ]);
+
+    const feeInfo = await this.#measureFeeInfo(
+      bundle,
+      bundleOverheadGas,
+    );
 
     return {
       feeDetected,
-      feeRequired,
+      feeRequired: feeInfo?.requiredFee ?? BigNumber.from(0),
       successes,
+      errorReason,
     };
   }
 
+  /**
+   * Removes rows that conflict with each other because they contain an
+   * operation for the same wallet and the same nonce.
+   */
+  async #filterRows(rows: BundleRow[]) {
+    const rowsByKeyAndNonce = new Map<string, BundleRow[]>();
+
+    for (const row of rows) {
+      for (let i = 0; i < row.bundle.operations.length; i++) {
+        const publicKey = row.bundle.senderPublicKeys[i];
+        const operation = row.bundle.operations[i];
+
+        const keyAndNonce = ethers.utils.solidityPack([
+          "uint256",
+          "uint256",
+          "uint256",
+          "uint256",
+          "uint256",
+        ], [...publicKey, operation.nonce]);
+
+        const entry = rowsByKeyAndNonce.get(keyAndNonce) ?? [];
+        entry.push(row);
+        rowsByKeyAndNonce.set(keyAndNonce, entry);
+      }
+    }
+
+    let filteredRows = rows;
+
+    for (let rowGroup of rowsByKeyAndNonce.values()) {
+      rowGroup = rowGroup.filter((r) => filteredRows.includes(r));
+
+      if (rowGroup.length <= 1) {
+        continue;
+      }
+
+      const bestRow = await this.#pickBest(rowGroup);
+      const conflictingRows = rowGroup.filter((r) => r !== bestRow);
+
+      filteredRows = filteredRows.filter((r) => !conflictingRows.includes(r));
+    }
+
+    return filteredRows;
+  }
+
+  async #pickBest(rows: BundleRow[]) {
```

Comment:
Could we add a comment here explaining how we pick the best?

Reply:
Fixed in https://github.com/web3well/bls-wallet/pull/438